newIDE/app/src/Profile/SubscriptionDialog.js

```diff
@@ -33,319 +29,333 @@ import {
-    if (!this.props.open && newProps.open) {
-      sendSubscriptionDialogShown();
-    }
-  }
+  React.useEffect(
+    () => {
+      if (open) {
+        sendSubscriptionDialogShown();
+      }
+    },
+    [open]
+  );
 
-  choosePlan = (
-    i18n: I18nType,
-    authenticatedUser: AuthenticatedUser,
-    plan: PlanDetails
-  ) => {
-    const {
-      getAuthorizationHeader,
-      subscription,
-      firebaseUser,
-    } = authenticatedUser;
-    if (!firebaseUser || !subscription) return;
+  const choosePlan = async (i18n: I18nType, plan: PlanDetails) => {
+    const { getAuthorizationHeader, subscription, profile } = authenticatedUser;
+    if (!profile || !subscription) return;
     sendChoosePlanClicked(plan.planId);
 
-    if (subscription.stripeSubscriptionId) {
-      const answer = Window.showConfirmDialog(
-        plan.planId
-          ? i18n._(t`Are you sure you want to subscribe to this new plan?`)
-          : i18n._(t`Are you sure you want to cancel your subscription?`)
-      );
-      if (!answer) return;
-
-      // We already have a stripe customer, change the subscription without
-      // asking for the user card.
-      this.setState({ isLoading: true });
-      changeUserSubscription(getAuthorizationHeader, firebaseUser.uid, {
-        planId: plan.planId,
-      }).then(
-        () =>
-          this.handleUpdatedSubscriptionSuccess(i18n, authenticatedUser, plan),
-        (err: Error) => this.handleUpdatedSubscriptionFailure(i18n, err)
-      );
-    } else {
-      this.setState({
-        subscriptionPendingDialogOpen: true,
-      });
+    // Subscribing from an account with a subscription.
+    if (
+      !subscription.stripeSubscriptionId &&
+      !subscription.paypalSubscriptionId
+    ) {
+      setSubscriptionPendingDialogOpen(true);
       Window.openExternalURL(
-        getRedirectToCheckoutUrl(
-          plan.planId || '',
-          firebaseUser.uid,
-          firebaseUser.email || ''
-        )
+        getRedirectToCheckoutUrl(plan.planId || '', profile.id, profile.email)
       );
+      return;
     }
-  };
 
-  handleUpdatedSubscriptionSuccess = (
-    i18n: I18nType,
-    authenticatedUser: AuthenticatedUser,
-    plan: PlanDetails
-  ) => {
-    authenticatedUser.onRefreshUserProfile();
-    this.setState({ isLoading: false });
-    if (plan.planId) {
-      showMessageBox(
-        i18n._(
-          t`Congratulations, your new subscription is now active!\n\nYou can now use the services unlocked with this plan.`
-        )
-      );
+    const { planId } = plan;
+    if (!planId) {
+      // Cancelling the existing subscription.
+      const answer = await showConfirmation(cancelConfirmationTexts);
+      if (!answer) return;
+
+      setIsChangingSubscription(true);
+      try {
+        await changeUserSubscription(getAuthorizationHeader, profile.id, {
+          planId: null,
+        });
+        await authenticatedUser.onRefreshUserProfile();
```

Comment:
I'm wondering if the onRefreshUserProfile() should be embedded in the `changeUserSubscription()` function, to avoid having to call it every time.
Not a big deal though

Reply:
It's an API call, so a bit hard to do it inside. I'll leave it as is for now!